lib/checkstl.cpp

```diff
@@ -1869,3 +1869,249 @@ void CheckStl::readingEmptyStlContainerError(const Token *tok, const ValueFlow::
+        return nullptr;
+
+    const Token * dotTok = start->tokAt(2);
+    if(!Token::findmatch(dotTok->tokAt(2), "%varid%", endStatement, varid))
+        return nullptr;
+    input = Token::Match(start->next(), "%var% . %name% ( %varid% )", varid);
+    if(isVariableChanged(dotTok->next(), endStatement, dotTok->astOperand1()->varId(), false, nullptr, true))
+        return nullptr;
+    return dotTok;      
+}
+
+static const Token* singleIncrementInScope(const Token* start, unsigned int varid, bool& input)
+{
+    if(start->str() != "{")
+        return nullptr;
+    const Token * endToken = start->link();
+    const Token * varTok = nullptr;
+    if(Token::Match(start->next(), "++ %var% ;"))
+        varTok = start->tokAt(2);
+    else if(Token::Match(start->next(), "%var% ++ ;"))
+        varTok = start->tokAt(1);
+    const Token * endStatement = start->tokAt(3);
+    if(endStatement->next() != endToken)
+        return nullptr;
+    
+    if(!varTok)
+        return nullptr;
+
+    input = varTok->varId() == varid;
+    return varTok;      
+}
+
+static const Token* singleConditionalInScope(const Token* start, unsigned int varid)
+{
+    if(start->str() != "{")
+        return nullptr;
+    const Token * endToken = start->link();
+    if(!Token::Match(start->next(), "if ("))
+        return nullptr;
+    if(!Token::Match(start->tokAt(2)->link(), ") {"))
+        return nullptr;
+    const Token * bodyTok = start->tokAt(2)->link()->next();
+    const Token * endBodyTok = bodyTok->link();
+    if(!Token::Match(endBodyTok, "} }"))
+        return nullptr;
+    if(endBodyTok->next() != endToken)
+        return nullptr;
+    if(!Token::findmatch(start, "%varid%", bodyTok, varid))
+        return nullptr;
+    if(isVariableChanged(start, bodyTok, varid, false, nullptr, true))
+        return nullptr;
+    return bodyTok;
+}
+
+static bool addByOne(const Token * tok, unsigned int varid)
+{
+    if(Token::Match(tok, "+= %any% ;") && 
+        tok->tokAt(1)->hasKnownIntValue() && 
+        tok->tokAt(1)->getValue(1)) {
+        return true;
+    }
+    if(Token::Match(tok, "= %varid% + %any% ;", varid) && 
+        tok->tokAt(3)->hasKnownIntValue() && 
+        tok->tokAt(3)->getValue(1)) {
+        return true;
+    }
+    return false;
+}
+
+void CheckStl::loopAlgo()
+{
+    if (!mSettings->isEnabled(Settings::STYLE))
+        return;
+    for (const Scope *function : mTokenizer->getSymbolDatabase()->functionScopes) {
+        for (const Token *tok = function->bodyStart; tok != function->bodyEnd; tok = tok->next()) {
+            // Parse range-based for loop
+            if(!Token::simpleMatch(tok, "for ("))
+                continue;
+            if(!Token::simpleMatch(tok->next()->link(), ") {"))
+                continue;
+            const Token * bodyTok = tok->next()->link()->next();
+            const Token * splitTok = Token::findsimplematch(tok->next(), ":", bodyTok);
+            if(!splitTok)
```

Comment:
I guess you should check if rhs is a container, that is simple with code like:

    const Token * rhs = splitTok->astOperand2();
    if (!rhs->valueType() || rhs->valueType()->type != ValueType:Type::CONTAINER)
        continue;

Reply:
It seems it would be to restrictive. A range-for loop only requires a range not a container, and we know its a range because its being used in a range-for loop.